integration/cluster_test.go

```diff
@@ -371,3 +374,47 @@ func clusterMustProgress(t *testing.T, membs []*member) {
 		mcancel()
 	}
 }
+
+func TestStopTriggerTransferLeader(t *testing.T) {
+	defer testutil.AfterTest(t)
+
+	clus := NewClusterV3(t, &ClusterConfig{Size: 3})
+	defer clus.Terminate(t)
+
+	leaderIdx1 := clus.waitLeader(t, clus.Members)
+	leaderID1 := clus.Members[leaderIdx1].s.ID()
+
+	donec := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.TODO())
+	go func() {
+		defer close(donec)
+
+		// fail when the leader is taken down without transfer
+		// and succeed if transfer works
+		for {
+			kvc := toGRPC(clus.Client((leaderIdx1 + 1) % 3)).KV
+			if _, err := kvc.Range(ctx, &etcdserverpb.RangeRequest{Key: []byte("foo")}); err != nil {
+				switch grpc.ErrorDesc(err) {
+				case context.Canceled.Error():
+					return
+				default:
+					t.Fatal(err)
+				}
+			}
+		}
+	}()
+
+	// stop leader
+	clus.Members[leaderIdx1].Stop(t)
+	<-clus.Members[leaderIdx1].s.StopNotify()
+
+	leaderIdx2 := clus.waitLeader(t, clus.Members)
+	cancel()
```

Comment:
Pretty sure this is still masking broken transfer. If the leader doesn't transfer smoothly then the range will eventually timeout, possibly after the cluster elects a new leader. In that case, the cancel will terminate the range request before it gets lost leader error response. The donec1/donec2 synchronization for shutting down the goroutine from before was correct.